pkgs/top-level/lua-packages.nix

```diff
@@ -99,6 +99,51 @@ with self; {
 
   luarocks-nix = callPackage ../development/tools/misc/luarocks/luarocks-nix.nix { };
 
+  luv = pkgs.stdenv.mkDerivation rec {
```

Comment:
this doesn't appear as a lua module anymore, thus installation of luaPackages.nvim-client fails. You can add toLuaModule() but it triggers another issue. Also it doesn't seem to install lua files anymore does it ?

Reply:
btw I've created this https://github.com/luvit/luv/issues/457

Reply:
do you need luarocks 3.3.1 ? I've reverted the luarocks bump in my fork, maybe we could do likewise until we find a better fix like the one mentioned in my luv issue.

Reply:
:+1: for catching that @teto (`neovim` builds fine because `doCheck = false` by default). After a while trying to debug this - I've finally realized that the `BUILD_MODULE` cmake flag is what's so problematic - they either install it as a module without headers (which is good only for native Lua packages) or they install everything (which is suitable for neovim). This very issue is what created the original motivation for the `libluv` and `luv` different derivations ! I've failed to understand that earlier..

Let's wait for their response for my https://github.com/luvit/luv/pull/458#issuecomment-591316547 .

Reply:
To be fair, there is also the problem that the current nixpkgs lua infrastructure doesn't work with cmake based project, I've worked quite a bit on it yesterday and found some issues with luarocks . I think I could solve it via creating a cmake wrapper but meanwhile I've proceeded to upload some unrelated changes I had in my fork https://github.com/NixOS/nixpkgs/pull/81067 . I don't have the space to run a nixpkgs-review on it (trying to buy a new SSD) but it should be ok overall.

Reply:
OK now that they've accepted our request and 1.34.2-0 got out, I tried updating once more luv and I got the tree I was hoping to get. Yet `nvim-client` still fails to build. Although `lua -e 'require("luv")'` doesn't fail as it's found in `LUA_CPATH`.

I tried using `buildLuaPackage` instead of `stdenv.mkDerivation` and it didn't help - I got the error:

```
error: attribute 'rocksSubdir' missing, at /var/src/nixpkgs/pkgs/development/interpreters/lua-5/build-lua-package.nix:121:86
(use '--show-trace' to show detailed location information)
```

When trying to build `nvim-client`.

I've also tried to revert back to using the generated `buildLuarocksPackage` and using the overrides. But it seems that this:

```
    extraVariables = ''
      WITH_SHARED_LIBUV="ON";
      LUA_BUILD_TYPE="System";
      BUILD_MODULE="ON";
      BUILD_SHARED_LIBS="ON";
      LUA_COMPAT53_DIR="${lua.pkgs.compat53}";
    '';
```
Doesn't help.

<details><summary><code>WITH_SHARED_LIBUV="ON"</code> isn't detected according to:</summary>

```
CMake Error at CMakeLists.txt:66 (add_subdirectory):
  add_subdirectory given source "deps/libuv" which is not an existing
  directory.


CMake Error at CMakeLists.txt:71 (set_target_properties):
  set_target_properties Can not find target to add properties to: uv_a
```

(While `deps` dir was removed).

</details>

I can see why https://github.com/luarocks/luarocks/issues/1160 is relevant now. There are other similar issues already open - https://github.com/luarocks/luarocks/issues/509 & https://github.com/luarocks/luarocks/issues/339

To my view a workaround for these issues on our end is missing..

I tried using the workaround that was used in the past where we've had an ugly `sed -i CMakeLists.txt` but that still fails for me now:

```
CMake Error at CMakeLists.txt:229 (install):
  install FILES given no DESTINATION!
```

Reply:
I had the same issue with `rocksSubdir` when I tried, I forgot to push the fix will do soon. I am glad you found out the issues about luarocks + cmake issues, since it means the issue I created is a duplicate but also a shared concern.
I believe for now, while waiting for proper luarocks + cmake support (either via a luarocks fix or we can hack a cmake wrapper with our cmakeFlags and have luarocks call it), we should aim at minimizing changes. Thus I encourage the approach of adjusting the sed -i as you did.
Looks to me that the error you have might be because SHAREDLIBS_INSTALL_INC_DIR is empty ?

Reply:
> Looks to me that the error you have might be because SHAREDLIBS_INSTALL_INC_DIR is empty ?

Correct, I tried to debug this from within luv's repo and I've failed to reproduce. I think I can narrow down the issue to here:

https://github.com/luvit/luv/blob/0938751421a1a1d73f5c98f1539a1f5a1c8d63e7/CMakeLists.txt#L203-L206

In comparison, the cmake variable `INSTALL_LIB_DIR` is set here: https://github.com/luvit/luv/blob/0938751421a1a1d73f5c98f1539a1f5a1c8d63e7/CMakeLists.txt#L86

I tried patching locally by adding an `INCDIR` (as an equivallent to `LIBDIR`) both here ^ and in the rockspec but it seems that `luarocks make` doesn't set this variable..

Reply:
Here's my attempted patch:

```
diff --git i/CMakeLists.txt w/CMakeLists.txt
index ed0f498..0c6571f 100644
--- i/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -84,6 +84,7 @@ if (LUA)
   SET(LUA_PACKAGE_CPATH "${LIBDIR}")
 
   SET(INSTALL_LIB_DIR ${LIBDIR})
+  SET(INSTALL_INC_DIR ${INCDIR})
 
   GET_FILENAME_COMPONENT(LUA_EXEC_NAME ${LUA_EXECUTABLE} NAME_WE)
   IF(LUA_EXEC_NAME STREQUAL "luajit")
diff --git i/luv-1.34.2-0.rockspec w/luv-1.34.2-0.rockspec
index 687039b..9c1d450 100644
--- i/luv-1.34.2-0.rockspec
+++ w/luv-1.34.2-0.rockspec
@@ -29,6 +29,7 @@ build = {
      LUA_INCDIR="$(LUA_INCDIR)",
      LUA="$(LUA)",
      LIBDIR="$(LIBDIR)",
+     INCDIR="$(INCDIR)",
      LUADIR="$(LUADIR)",
   },
 }
```

Reply:
But, this patch does seem to work a little bit better:

```
diff --git i/CMakeLists.txt w/CMakeLists.txt
index ed0f498..0c6571f 100644
--- i/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -84,6 +84,7 @@ if (LUA)
   SET(LUA_PACKAGE_CPATH "${LIBDIR}")
 
   SET(INSTALL_LIB_DIR ${LIBDIR})
+  SET(INSTALL_INC_DIR "${LIBDIR}/../include")
 
   GET_FILENAME_COMPONENT(LUA_EXEC_NAME ${LUA_EXECUTABLE} NAME_WE)
   IF(LUA_EXEC_NAME STREQUAL "luajit")
```

Yet it fails with the same errors as before I dived into fixing the whole thing:

```
Install the project...
-- Install configuration: ""
-- Installing: /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/luv.so
-- Installing: /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/libluv.so.1.34.2
-- Installing: /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/libluv.so.1
-- Installing: /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/libluv.so
-- Installing: /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/pkgconfig/libluv.pc
-- Installing: /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/../include/luv.h
-- Installing: /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/../include/util.h
-- Installing: /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/../include/lhandle.h
-- Installing: /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/../include/lreq.h
cp: cannot stat '/nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/libluv.so': No such file or directory

Error: Failed copying /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib/libluv.so to /nix/store/c9pbzvmjgg5kgpwqq376wg9mbafkhsl6-luajit-2.1.0-beta3-luv-1.34.2-0/lib/lua/5.1/libluv.so
```

I wonder now whether `luarocks make` doesn't call cmake properly - whether I should have investigate better and reported a little bit better this: https://github.com/luarocks/luarocks/issues/1155

Reply:
the level of indirection just to set some SHAREDLIBS_INSTALL_INC_DIR is kinda confusing, maybe report it to luv ?

Reply:
I'm not sure what do you mean by "indirection". But In any case, what ever complaints you have, ironically _I_ am the person to blame because I wrote the fix for https://github.com/luvit/luv/issues/457 which introduced these new cmake variables :).

I'm mostly perplexed as for what's going on when `luarocks make` runs - that's the real, original source of our issue. It seems to be acting rather differently then just a plain `cmake .. && make`. To my opinion, debugging this and waiting for this behavior to be fixed up there at `luarocks` is going to be hard. Hence It seems to me we are left with 2 options:

- Stay with the current scheme of building this project with `stdenv.mkDerivation` and _somehow_[1] resolve the issue of `nvim-client` not finding `luv`.
- Use `buildLuarocksPackage` (so that `nvim-client` will detect `luv` as a dependency) and completely replace / override `luv`'s phases in order to avoid the original error that required this PR.

[1]: @teto as the main maintainer of the Lua ecosystem in Nixpkgs, do you have any idea for a workaround to make `nvim-client` find `luv`? It seems related to it's setup-hook but I'm not sure exactly where to look for.

Reply:
haha sry for that, SHAREDLIBS_INSTALL_INC_DIR comes from either INSTALL_INC_DIR (which seems never defined ?) or ${CMAKE_INSTALL_PREFIX}/include/luv https://github.com, all of those in if/else. `luarocks make` is pretty straightforward, it just calls cmake with the `variables` defined in the rockspec see  /luarocks/luarocks/blob/master/src/luarocks/build/cmake.lua . The best way IMO (fastest/cleanest) is to keep nixpkgs as is (or almost) but just let luv correctly install the library. Improving luv's CMakeList readibility concerning SHAREDLIBS_INSTALL_INC_DIR would be a good idea. Thanks for the help so far.

Reply:
I see. Then the issue is that `luarocks make` doesn't play nice with `luv`'s current `CMakeLists.txt`. You've convinced me that we should contact them once more, but I'm not sure how to explain this to them..

Reply:
luv should make sure that SHAREDLIBS_INSTALL_INC_DIR is not empty, it should be possible to define it with the existing variables.

Reply:
Agree. I'm working on a fix for luv's CMakeLists right now, feel free to open an issue first there and explain the issue.

Reply:
@teto while investigating the issue with luv's CMakeLists, I've come to the conclusion that `luarocks make` sets too many cmake variables, and I don't think we can stop it from doing it. Here's the list of them extracted out of a `nix-build -A luajit.pkgs.luv` while `NIX_DEBUG = 2;` was set in luv's derivation:

```
-G"Unix Makefiles"
-DCMAKE_MODULE_LINKER_FLAGS="-shared"
-DLIBDIR="/nix/store/5wf8zrgpp199pmbbgyvkp7hc484pzrbs-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34.2-0-rocks/luv/1.34.2-0/lib"
-DLUA_INCDIR="/nix/store/1m8vh1byjnjmjkpdidcqc37cykgsqjap-luajit-2.1.0-beta3/include/luajit-2.1"
-DLUA_LIBDIR=""
-DLUA="/nix/store/1m8vh1byjnjmjkpdidcqc37cykgsqjap-luajit-2.1.0-beta3/bin/luajit"
-DCMAKE_LIBRARY_PATH="/nix/store/1m8vh1byjnjmjkpdidcqc37cykgsqjap-luajit-2.1.0-beta3/lib:/nix/store/w89ndqsb7ig22mw5hk7k7zfvxmhx44xb-libuv-1.34.2/lib:/nix/store/1m8vh1byjnjmjkpdidcqc37cykgsqjap-luajit-2.1.0-beta3/lib:/nix/store/w89ndqsb7ig22mw5hk7k7zfvxmhx44xb-libuv-1.34.2/lib"
-DCMAKE_INCLUDE_PATH="/nix/store/1m8vh1byjnjmjkpdidcqc37cykgsqjap-luajit-2.1.0-beta3/include:/nix/store/w89ndqsb7ig22mw5hk7k7zfvxmhx44xb-libuv-1.34.2/include:/nix/store/1m8vh1byjnjmjkpdidcqc37cykgsqjap-luajit-2.1.0-beta3/include:/nix/store/w89ndqsb7ig22mw5hk7k7zfvxmhx44xb-libuv-1.34.2/include"
-DLUADIR="/nix/store/5wf8zrgpp199pmbbgyvkp7hc484pzrbs-luajit-2.1.0-beta3-luv-1.34.2-0/luv-1.34"
```

There are a couple of troubles with this:

1. `LIBDIR` is set to some distant path inside `$out` and not just `$out/lib`.
2. `LUA_LIBDIR` is empty.

In my opinion, (1) is severe and unfixable - it's hard-coded inside `luarocks make`. I think `luarocks make` can't even imagine itself running a cmake command that will imply installing files in a `$prefix/lib` instead of `$prefix/${rockname-version}-rocks/${rockname}/${version}/lib`. Even if we'll fix the original issue `cp` error this PR was written for, I doubt that `neovim` will be able to find pkg-config files for luv inside the standard `${luv}/lib` path.

I'm not sure what's the effects of (2) but I don't like it either.

This is very messy indeed, most likely because `luarocks` doesn't expect packages that provide headers and shared libraries someone (or a project like neovim) would expect to find in the standard locations. And I think I realize now that according to this behaviour, the original design of luv's CMakeLists files chose to make the BUILD_MODULE and BUILD_SHARED_LIBS colliding.

While writing this, I think perhaps we should provide a `libluv` package in `lua-packages.nix` that will use `stdenv.mkDerivation` and that will be used by neovim along with another `luv` package that we'll only override, just as before.

Reply:
I've tried my idea and I think it's the most elegant solution to the issue. Especially considering `luarocks` lack of support for installing arbitrary files to a `$prefix`, as explained above. I've pushed the final fix for this.

Reply:
having to maintain the same package twice because its build system cant find where to install files does not seem elegant + it will keep being a headache; better fix it now that we have everything in mind. As for the cmake call via luarocks; it s not a problem; you can override the parameters via environment variables or patch the luv rockspec. We better open an issue on luv. I am busy tomorrow but will try to do at least saturday.